### PR TITLE
[TT-1490] add gas bumping

### PR DESCRIPTION
**Pull request by Tofel**
this PR adds gas bumping for transactions that are slow to confirm. that means that they need to be successfully submitted and cannot be reverted. by default bumping is disabled (`retries = 0`). there's also an option to specify max gas price in wei.

gas bumping can be used, when deploying contracts via `DeployContract()` function and inside `Decode()` function, when waiting for tx to be mined.

following tests were added:
- [x] (legacy tx) contract deployment: successful bumping with standard function
- [x] (legacy tx) contract deployment: bumping disabled
- [x] (legacy tx) contract deployment: failed bumping
- [x] (legacy tx) contract deployment: successful bumping with custom function
- [x] (dynamic tx) contract deployment: successful bumping with standard function
- [x] (dynamic tx) contract deployment: bumping of tx from non-root key
- [x] (dynamic tx) contract deployment: bumping of tx from unknown/invalid key
- [x] (legacy tx) contract interaction: successful bumping with standard function
- [x] (legacy tx) contract interaction: bumping disabled
- [x] (legacy tx) contract interaction: failed bumping

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -548,6 +551,94 @@ Here's what they do:
...
+
+Once the gas price bump would go above the limit we stop bumping and use the last gas price that was below the limit.
+
+How gas price is calculated depends on transaction type:
+- for legacy transactions it's just the gas price
+- for EIP-1559 transactions it's the sum of gas fee cap and tip cap
+- for Blob transactions (EIP-4844) it's the sum of gas fee cap and tip cap and max fee per blob
+- for AccessList transactions (EIP-2930) it's just the gas price
+
+Please note that Blob and AccessList
```
Comment (skudasov) [inline README.md:581]: Should we end the sentence? =)

File: README.md [hunk 2]
```diff
@@ -548,6 +551,94 @@ Here's what they do:
...
+    if suggestedGasPrice.Cmp(gasPrice) == 1 {
+        return suggestedGasPrice
+    }
+
+	// otherwise bump by 100%
+    return new(big.Int).Mul(gasPrice, big.NewInt(2))
+}
+```
+
+Currently, the gas bumping mechanism is available for:
```
Comment (skudasov) [inline README.md:623]: Let's remove these lines, we've already described them before, or shorten it

File: client.go [hunk 3]
```diff
@@ -439,7 +452,42 @@ func (m *Client) Decode(tx *types.Transaction, txErr error) (*DecodedTransaction
...
+				return
+			} else {
+				L.Debug().Str("Current error", retryErr.Error()).Uint("Attempt", i).Msg("Waiting for transaction to be confirmed after gas bump")
+			}
+			tx = replacementTx
+		}),
+		retry.DelayType(retry.FixedDelay),
+		// unless attempts is at least 1 retry.Do won't execute at all
+		retry.Attempts(func() uint {
+			if m.Cfg.GasBumpRetries() == 0 {
```
Comment (skudasov) [inline client.go:480]: Is it strictly 1 attempt still or 2 attempts, 1 initial and 1 after?

File: README.md [hunk 4]
```diff
@@ -548,6 +551,94 @@ Here's what they do:
...
+When enabled, gas bumping is used in two places:
+* during contract deployment via `DeployContract` function
+* inside `Decode()` function
+
+It is recommended to decrease transaction timeout when using gas bumping, as it will be effectively increased by the number of retries. So if you were running with 5 minutes timeout and 0 retries, you should set it to 1 minute and 5 retries
+or 30 seconds and 10 retries.
+
+Don't worry if while bumping logic executes previous transaction gets mined. In that case sending replacement transaction with higher gas will fail (because it is using the same nonce as original transaction) and we will retry waiting for the mining of the original transaction.
+
+**Gas bumping is only applied for submitted transaction. If transaction was rejected by the node (e.g. because of too low base fee) we will not bump the gas price nor try to submit it, because original transaction submission happens outside of Seth.**
```
Comment (skudasov) [inline README.md:640]: Sadly enough, that's true, haven't found any other possible workarounds?
`transact` in generated bindings returns `nil` if it fails to send
```
	if err := c.transactor.SendTransaction(ensureContext(opts.Context), signedTx); err != nil {
		return nil, err
	}
	return signedTx, nil
```

File: retry.go [hunk 5]
```diff
@@ -47,3 +51,198 @@ func (m *Client) RetryTxAndDecode(f func() (*types.Transaction, error)) (*Decode
...
+	case Priority_Degen:
+		// +100%
+		return func(gasPrice *big.Int) *big.Int {
+			return gasPrice.Mul(gasPrice, big.NewInt(2))
+		}
+	case Priority_Fast:
+		// +30%
+		return func(gasPrice *big.Int) *big.Int {
+			gasPriceFloat, _ := gasPrice.Float64()
+			newGasPriceFloat := big.NewFloat(0.0).Mul(big.NewFloat(gasPriceFloat), big.NewFloat(1.5))
```
Comment (skudasov) [inline retry.go:76]: Did you mean 1.3 for 30%?

**Review by skudasov — CHANGES_REQUESTED** [review]

File: client.go (hunk 3 above)
Comment (Tofel) [reply client.go:480]: that's the thing, if attempts is `0` no retry will ever be done

File: README.md (hunk 4 above)
Comment (Tofel) [reply README.md:640]: yup so inside `Decode()` we don't have the transaction so we cannot resubmit it, the only solution I can think of is what I described to you yesterday: 
* implement `Backend`
* proxy `SendTransaction` to original one
* if err, retry inside our implementation of `SendTransaction`

File: retry.go (hunk 5 above)
Comment (Tofel) [reply retry.go:76]: good catch!

**Review by skudasov — APPROVED** [review]

**Pull request merged**